plugins/c8y_firmware_plugin/src/firmware_manager.rs

```diff
@@ -0,0 +1,197 @@
+            TopicFilter::new_unchecked(FIRMWARE_UPDATE_RESPONSE_TOPICS);
+
+        let firmware_download_manager = FirmwareDownloadManager::new(
+            tedge_device_id,
+            mqtt_client.published.clone(),
+            http_client.clone(),
+            local_http_host,
+            tmp_dir,
+            timeout_sec,
+        );
+
+        Ok(FirmwareManager {
+            mqtt_client,
+            c8y_request_topics,
+            health_check_topics,
+            firmware_update_response_topics,
+            firmware_download_manager,
+        })
+    }
+
+    pub async fn run(&mut self) -> Result<(), anyhow::Error> {
+        self.get_pending_operations_from_cloud().await?;
+
+        // Now the firmware plugin is done with the initialization and ready for processing the messages
+        send_health_status(&mut self.mqtt_client.published, PLUGIN_SERVICE_NAME).await;
+        info!("Ready to serve the firmware request.");
+
+        loop {
+            tokio::select! {
+                message = self.mqtt_client.received.next() => {
+                    if let Some(message) = message {
+                        let topic = message.topic.name.clone();
+                        if let Err(err) = self.process_mqtt_message(
+                            message,
+                        )
+                        .await {
+                            error!("Processing the message received on {topic} failed with {err}");
+                        }
+                    } else {
+                        // message is None and the connection has been closed
+                        return Ok(())
+                    }
+                }
+                Some(((child_id, hash), op_state)) = self.firmware_download_manager.operation_timer.next_timed_out_entry() => {
+                    let failure_reason = format!("Child device {child_id} did not respond within the timeout interval of {}sec. Timer ID={hash}",
+                        self.firmware_download_manager.timeout_sec.as_secs());
+                    info!(failure_reason);
+                    mark_pending_firmware_operation_failed(self.mqtt_client.published.clone(), &child_id, op_state,failure_reason).await?;
+                }
+            }
+        }
+    }
+
+    async fn process_mqtt_message(&mut self, message: Message) -> Result<(), anyhow::Error> {
+        if self.health_check_topics.accept(&message) {
+            send_health_status(&mut self.mqtt_client.published, PLUGIN_SERVICE_NAME).await;
+            return Ok(());
+        } else if self.firmware_update_response_topics.accept(&message) {
+            self.handle_child_device_firmware_operation_response(&message)
+                .await?
+        } else if self.c8y_request_topics.accept(&message) {
+            for smartrest_message in collect_smartrest_messages(message.payload_str()?) {
+                let result = match smartrest_message.split(',').next().unwrap_or_default() {
```

Comment:
```suggestion
                let result = c8y_api::smartrest::message::get_smartrest_template_id(smartrest_message) {
```
An API exists for this already.